src/download.c

```diff
@@ -125,23 +115,34 @@ static int swupd_curl_hashmap_insert(struct file *file)
 	unlink(tar_dotfile);
 	free_string(&tar_dotfile);
 
-	// queue the hash for download
-	iter = *bucket;
-	if ((iter = list_prepend_data(iter, file)) == NULL) {
-		return -1;
-	}
-	*bucket = iter;
-
 	return 0;
 }
 
 /* hysteresis thresholds */
 static size_t MAX_XFER = 25;
 static size_t MAX_XFER_BOTTOM = 15;
 
+static bool file_hash_cmp(const void *a, const void *b)
+{
+	const struct file *fa = a;
+	const struct file *fb = b;
+
+	return hash_equal(fa->hash, fb->hash);
+}
+
+static size_t file_hash_value(const void *data)
+{
+	return HASH_VALUE(((struct file *)data)->hash[0]);
+}
+
 int start_full_download(bool pipelining)
 {
+	if (swupd_curl_hashmap) {
+		return -1;
+	}
+
 	failed = NULL;
+	swupd_curl_hashmap = hashmap_new(SWUPD_CURL_HASH_BUCKETS, file_hash_cmp, file_hash_value);
```

Comment:
`hashmap_new` can return `NULL` if the `calloc` fails. Are you sure all code that uses `swupd_curl_hashmap` after this point is checking for this case?